File: c++/Mandelbrot/mandelbrot.cpp
#include <iostream>
#include <utility>
#include <vector>
#include <complex>
#include <cmath>
#include <sstream>
#include "stats.hpp"

#ifdef SAVE_IMAGE
    #include <SFML/Graphics.hpp>
#endif

// black or white
enum class Color : bool {
    WHITE = true,
    BLACK = false
};
constexpr Color inSetColor = Color::WHITE;
constexpr Color notInSetColor = Color::BLACK;

typedef double Real;

typedef std::complex<Real> Complex;

std::ostream& operator<<(std::ostream& out, Complex const& z)
{
    return out << "(" << z.real() << ";" << z.imag() << ")";
}

typedef std::pair<Complex, Complex> ComplexRange;

std::ostream& operator<<(std::ostream& out, ComplexRange const& range)
{
    return out << "{" << range.first << ";" << range.second << "}";
}

typedef std::vector<Color> Image; // square image stored in a 1D array

std::size_t getSideSize(Image const& img)
{
    return std::sqrt(img.size());
}

Color& getPixel(Image& img, std::size_t x, std::size_t y)
{
    return img[y * getSideSize(img) + x];
}

struct Mandelbrot
{
    Mandelbrot(std::size_t side, std::size_t maxIterations, ComplexRange range)
    : side(side)
    , maxIterations(maxIterations)
    , range(range)
    { /* - */ }

    // Perform the set computation
    void operator()() const
    {
        Image img(side * side);
        for (std::size_t x = 0; x < side; ++x) {
            for (std::size_t y = 0; y < side; ++y) {
                getPixel(img, x, y) = computeElement(x, y);
            }
        }

        #ifdef SAVE_IMAGE
        static std::size_t imgId = 0;
        sf::Image png; png.create(side, side, sf::Color::White);

        for (std::size_t x = 0; x < side; ++x) {
            for (std::size_t y = 0; y < side; ++y) {
                png.setPixel(x, y, getPixel(img, x, y) == inSetColor ? sf::Color::Black : sf::Color::White);
            }
        }

        png.saveToFile("tmp/fractal_" + std::to_string(imgId) + "_" + csvdescription() + ".png");
        ++imgId;
        #endif
    }
    
    std::string csvdescription() const 
    {
        std::stringstream ss;
        ss << side << "," 
           << maxIterations << "," 
           << range;
        return ss.str();
    }

    // Compute the color of one element of the set
    Color computeElement(std::size_t x, std::size_t y) const
    {
        const Complex c {
            range.first.real() + x / (side - 1.0) * (range.second.real() - range.first.real()),
            range.first.imag() + y / (side - 1.0) * (range.second.imag() - range.first.imag())
        };

        Complex z = { 0, 0 };

        std::size_t iter = 0;
        for (iter = 0; iter < maxIterations && std::abs(z) < 2.0; ++iter) {
            z = z * z + c;
        }

        return iter == maxIterations ? inSetColor : notInSetColor;
    }

    std::size_t side, maxIterations;
    ComplexRange range;
};


int main(int, const char**)
{
    std::size_t maxIterations = 1000;
    std::size_t side = 2000;
    ComplexRange range = { Complex(-1.72, 1.2), Complex(1.0, -1.2) };
    stats<Mandelbrot, void>(Mandelbrot(side, maxIterations, range), 1);

    return 0;
}

File: scala/Mandelbrot/src/main/scala/Mandelbrot.scala
import java.awt.image.{ BufferedImage }
import java.io.{ File }
import javax.imageio.{ ImageIO }
import scala.collection.workstealing._

case class ComplexRange(val first_r: Double, val first_i: Double, val second_r: Double, val second_i: Double) {
    override def toString(): String =  "{ (" + first_r + ";" + first_i + ") ; (" + second_r + ";" + second_i + ") }"  
}

case class Mandelbrot(val width: Int, val height: Int, 
                      val range: ComplexRange, val maxIteration: Int, 
                      val inSet: Int, val notInSet: Int) {
    def computeElement(index: Int): Int = {
        val x = index % width
        val y = index / height

        val c_r = range.first_r + x / (width - 1.0) * (range.second_r - range.first_r)
        val c_i = range.first_i + y / (height - 1.0) * (range.second_i - range.first_i)

        var z_r, z_i = 0.0

        var iter = 0
        while (iter < maxIteration && z_r * z_r + z_i * z_i < 4.0) {
            // z = z * z + c
            val tmp = z_r;
            z_r = z_r * z_r - z_i * z_i + c_r
            z_i = 2 * tmp * z_i + c_i
            iter = iter + 1
        }

        if (iter == maxIteration) inSet
        else notInSet
    }
}

object Mandelbrot {

    def main(args: Array[String]): Unit = {

        val sides = List( 100, 400, 1200, 2000, 4000, 10000 )
        val iterations = List( 1, 30, 250, 1000, 2000, 8000 )
        val ranges = List(
            ComplexRange( -1.72, 1.2, 1.0, -1.2 ),
            ComplexRange( -0.7, 0, 0.3, -1 ),
            ComplexRange( -0.4, -0.5, 0.1, -1 ),
            ComplexRange( -0.4, -0.6, -0.2, -0.8 ),
            ComplexRange( -0.24, -0.64, -0.26, -0.66 )
        )

        val inSet = 0x000000
        val notInSet = 0xffffff

        var imgId = 0;
        
        val parlevel = sys.props("par").toInt
        val fj = new collection.parallel.ForkJoinTaskSupport(new scala.concurrent.forkjoin.ForkJoinPool(parlevel))

        type ComputeFn = (Int, Int, ComplexRange) => Array[Int]
        
        def stats(side: Int, maxIteration: Int, range: ComplexRange, title: String, compute: ComputeFn) {

            // Warmup
            if (side <= 500 || side <= 1000 && maxIteration < 2000) {
                for (i <- 0 until 20) {
                    compute(side, maxIteration, range)
                }
            }

            // Clean the VM
            System.gc()

            // Start timer
            val tic = System.nanoTime()

            //// COMPUTATION STARTS HERE

            val img = compute(side, maxIteration, range)

            //// COMPUTATION ENDS HERE
            
            // Stop timer
            val toc = System.nanoTime();

            val µs = (toc - tic) / 1000

            // Display the result
            val csvdescription = "parallel" + "," + side + "," + maxIteration + "," + range
            println(csvdescription + "," + µs)

            // Export the set to PNG
            val png = new BufferedImage(side, side, BufferedImage.TYPE_INT_RGB)
            val indexes = 0 until (side * side)
            for((index, color)  <- indexes zip img) {
                val x = index % side
                val y = index / side

                png.setRGB(x, y, color)
            }

            val outputfile = new File("tmp/fractal_" + imgId + "_" + csvdescription + ".png");
            ImageIO.write(png, "png", outputfile);

            imgId += 1
        }
        
        // Generate an image of the Mandelbrot set using the current parallel collection
        def computeNormal(side: Int, maxIteration: Int, range: ComplexRange): Array[Int] = {
            val indexes = (0 until side * side).par
            indexes.tasksupport = fj

            val generator = Mandelbrot(side, side, range, maxIteration, inSet, notInSet)
            val img = Array.ofDim[Int](side * side) 
            indexes foreach { idx =>
              img(idx) = generator.computeElement(idx)
            }

            img
        }
        
        // Generate an image of the Mandelbrot set using the next generation of parallel collection
        def computeNewPC(side: Int, maxIteration: Int, range: ComplexRange): Array[Int] = {
            val indexes = new ParRange(0 until (side * side), Workstealing.DefaultConfig)

            val generator = Mandelbrot(side, side, range, maxIteration, inSet, notInSet)
            val img = Array.ofDim[Int](side * side) 
            indexes foreach { idx =>
              img(idx) = generator.computeElement(idx)
            }

            img
        }

        for {
          side <- sides
          maxIteration <- iterations
          range <- ranges
        }{
            stats(side, maxIteration, range, "normal", computeNormal)
            stats(side, maxIteration, range, "newpc", computeNewPC)
        }
    }
}

